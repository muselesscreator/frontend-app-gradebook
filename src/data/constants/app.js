import { StrictDict } from 'utils';

export const modalFieldKeys = StrictDict({
  adjustedGradePossible: 'adjustedGradePossible',
  adjustedGradeValue: 'adjustedGradeValue',
  assignmentName: 'assignmentName',
  reasonForChange: 'reasonForChange',
  todaysDate: 'todaysDate',
  updateModuleId: 'updateModuleId',
  updateUserId: 'updateUserId',
  updateUserName: 'updateUserName',
  open: 'open',
});

export const localFilterKeys = StrictDict({
  assignmentGradeMax: 'assignmentGradeMax',
  assignmentGradeMin: 'assignmentGradeMin',
  courseGradeMax: 'courseGradeMax',
  courseGradeMin: 'courseGradeMin',
});
