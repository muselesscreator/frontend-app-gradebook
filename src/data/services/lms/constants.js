import { StrictDict } from 'utils';

export const pageSize = 25;
export const historyRecordLimit = 5;

export const paramKeys = StrictDict({
  cohortId: 'cohort_id',
  pageSize: 'page_size',
  userContains: 'user_contains',
  enrollmentMode: 'enrollment_mode',
  assignment: 'assignment',
  assignmentGradeMin: 'assignmentGradeMin',
  assignmentGradeMax: 'assignmentGradeMax',
  courseGradeMin: 'courseGradeMin',
  courseGradeMax: 'courseGradeMax',
  excludedCourseRoles: 'excluded_course_roles',
});
