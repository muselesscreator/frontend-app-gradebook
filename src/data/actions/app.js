import { StrictDict } from 'utils';
import { modalFieldKeys, localFilterKeys } from 'data/constants/app';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

const closeModal = createAction('closeModal');
/**
 * setCourseId(courseId)
 * loads courseID into local app state
 * @param {string} courseId - course ID from app context
 */
const setCourseId = createAction('setCourseId');
/**
  * setModalStateFromTable({ userEntry, subsection })
  * sets modalState
  * */
const setModalStateFromTable = createAction('setModalStateFromTable');
/**
 * setSearchValue(searchValue)
 * sets searchValue in local state
 * @param {string} searchValue - search input text
 */
const setSearchValue = createAction('setSearchValue');

const filterReducer = (payload) => (obj, key) => (
  (payload[key] !== undefined) ? { ...obj, [key]: payload[key] } : obj
);

/**
 * setLocalFilter(filterObject)
 * sets a number of localFilterKey fields on localFilters
 * @param {object} filterObject - an object of filter values
 */
const setLocalFilter = createAction('setLocalFilter', (filterObject) => ({
  payload: Object.keys(localFilterKeys).reduce(filterReducer(filterObject), {}),
}));

/**
 * setModalState(modalState)
 * sets a number of modalFieldKey fields on modalState
 * @param {object} modalState - an object of modal state values
 */
const setModalState = createAction('setModalState', (modalState) => ({
  payload: Object.keys(modalFieldKeys).reduce(filterReducer(modalState), {}),
}));

export default StrictDict({
  closeModal,
  setCourseId,
  setModalState,
  setSearchValue,
  setLocalFilter,
  setModalStateFromTable,
});
