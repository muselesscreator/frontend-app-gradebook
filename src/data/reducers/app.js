import initialFilters from '../constants/filters';
import { formatDateForDisplay } from '../actions/utils';
import actions from '../actions/app';
import filterActions from '../actions/filters';

const initialState = {
  courseId: '',
  filters: {
    assignmentGradeMax: initialFilters.assignmentGradeMax,
    assignmentGradeMin: initialFilters.assignmentGradeMin,
    courseGradeMax: initialFilters.courseGradeMax,
    courseGradeMin: initialFilters.courseGradeMin,
  },
  modalState: {
    open: false,
    adjustedGradePossible: '',
    adjustedGradeValue: 0,
    assignmentName: '',
    reasonForChange: '',
    todaysDate: '',
    updateModuleId: null,
    updateUserId: null,
    updateUserName: null,
  },
  searchValue: '',
};

const app = (state = initialState, { type, payload }) => {
  switch (type) {
    case actions.closeModal.toString():
      return { ...state, modalState: { ...initialState.modalState } };
    case actions.setCourseId.toString():
      return { ...state, courseId: payload };
    case actions.setLocalFilter.toString():
      return {
        ...state,
        filters: { ...state.filters, ...payload },
      };
    case actions.setModalState.toString():
      return {
        ...state,
        modalState: { ...state.modalState, ...payload },
      };
    case actions.setModalStateFromTable.toString(): {
      const { subsection, userEntry } = payload;
      return {
        ...state,
        modalState: {
          adjustedGradeValue: '',
          open: true,
          reasonForChange: '',
          todaysDate: formatDateForDisplay(new Date()),

          adjustedGradePossible: subsection.attempted ? subsection.score_possible : '',
          assignmentName: `${subsection.subsection_name}`,
          updateModuleId: subsection.module_id,
          updateUserId: userEntry.user_id,
          updateUserName: userEntry.username,
        },
      };
    }
    case actions.setSearchValue.toString():
      return { ...state, searchValue: payload };
    // initialize the filter fields that are locally stored
    case filterActions.initialize.toString():
      return {
        ...state,
        filters: {
          assignmentGradeMax: payload.assignmentGradeMax,
          assignmentGradeMin: payload.assignmentGradeMin,
          courseGradeMax: payload.courseGradeMax,
          courseGradeMin: payload.courseGradeMin,
        },
      };
    // Reset only the filter fields that are stored locally
    case filterActions.reset.toString(): {
      return payload.filter(
        (filterName) => initialState.filters[filterName] !== undefined,
      ).reduce((obj, filterName) => ({
        ...obj,
        filters: {
          ...obj.filters,
          [filterName]: initialFilters[filterName],
        },
      }), { ...state });
    }
    default:
      return state;
  }
};

export { initialState };
export default app;
