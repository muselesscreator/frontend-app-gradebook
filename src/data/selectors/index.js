/* eslint-disable import/no-named-as-default-member, import/no-self-import */
import { StrictDict } from 'utils';
import LmsApiService from 'data/services/LmsApiService';

import * as module from '.';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import special from './special';
import tracks from './tracks';

/**
 * lmsApiServiceArgs(state)
 * Returns common lms api service request args.
 * @param {object} state - redux state
 * @return {object} lms api query params object
 */
export const lmsApiServiceArgs = (state) => ({
  cohort: cohorts.getCohortNameById(state, filters.cohort(state)),
  assignment: filters.selectedAssignmentId(state),
  assignmentType: filters.assignmentType(state),
  assignmentGradeMin: grades.formatMinAssignmentGrade(
    filters.assignmentGradeMin(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  assignmentGradeMax: grades.formatMaxAssignmentGrade(
    filters.assignmentGradeMax(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  courseGradeMin: grades.formatMinCourseGrade(filters.courseGradeMin(state)),
  courseGradeMax: grades.formatMaxCourseGrade(filters.courseGradeMax(state)),
});

/**
 * gradeExportUrl(state, options)
 * Returns the output of getGradeExportCsvUrl, applying the current includeCourseRoleMembers
 * filter.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {string} - generated grade export url
 */
export const gradeExportUrl = (state, { courseId }) => (
  LmsApiService.getGradeExportCsvUrl(courseId, {
    ...module.lmsApiServiceArgs(state),
    excludeCourseRoles: filters.includeCourseRoleMembers(state) ? '' : 'all',
  })
);

/**
 * interventionExportUrl(state, options)
 * Returns the output of getInterventionExportUrl.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {string} - generated intervention export url
 */
export const interventionExportUrl = (state, { courseId }) => (
  LmsApiService.getInterventionExportCsvUrl(
    courseId,
    module.lmsApiServiceArgs(state),
  )
);

/**
 * getHeadings(state)
 * Returns the table headings given the current assignmentType and assignmentLabel filters.
 * @param {object} state - redux state
 * @return {string[]} - array of table headings
 */
export const getHeadings = (state) => grades.headingMapper(
  filters.assignmentType(state) || 'All',
  filters.selectedAssignmentLabel(state) || 'All',
)(grades.getExampleSectionBreakdown(state));

/**
 * showBulkManagement(state, options)
 * Returns true iff the user has special access or bulk management is configured to be available
 * and the course has a masters track.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state, { courseId }) => (
  special.hasSpecialBulkManagementAccess(courseId)
  || (tracks.stateHasMastersTrack(state) && state.config.bulkManagementAvailable)
);

/**
 * shouldShowSpinner(state)
 * Returns true iff the user can view the gradebook and grades.showSpinner is true.
 * @param {object} state - redux state
 * @return {bool} - should show spinner?
 */
export const shouldShowSpinner = (state) => (
  roles.canUserViewGradebook(state)
  && grades.showSpinner(state)
);

/**
 * showBulkManagement(state, options)
 * Returns true iff the user has special access or bulk management is configured to be available
 * and the course has a masters track.
 * @param {object} state - redux state
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state) => (
  special.hasSpecialBulkManagementAccess(app.courseId(state))
  || (tracks.stateHasMastersTrack(state) && state.config.bulkManagementAvailable)
);

export default StrictDict({
  root: StrictDict({
    getHeadings,
    gradeExportUrl,
    interventionExportUrl,
    shouldShowSpinner,
  }),
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  special,
  tracks,
});
