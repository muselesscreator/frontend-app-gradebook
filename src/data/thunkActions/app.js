import { StrictDict } from 'utils';

import actions from 'data/actions';
import { fetchGradeOverrideHistory } from './grades';
import { fetchRoles } from './roles';

export const setModalStateFromTable = ({ userEntry, subsection }) => (
  (dispatch) => {
    dispatch(fetchGradeOverrideHistory(subsection.module_id, userEntry.user_id));
    dispatch(actions.app.setModalStateFromTable({ subsection, userEntry }));
  }
);

export const initialize = (courseId, urlQuery) => (dispatch) => {
  dispatch(actions.app.setCourseId(courseId));
  dispatch(actions.filters.initialize(urlQuery));
  dispatch(fetchRoles());
};

export default StrictDict({
  initialize,
  setModalStateFromTable,
});
