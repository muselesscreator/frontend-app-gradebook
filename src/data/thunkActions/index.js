import { StrictDict } from 'utils';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
});
