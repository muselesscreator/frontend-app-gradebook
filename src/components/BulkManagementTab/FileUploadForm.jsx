/* eslint-disable react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  Button,
  Form,
  FormControl,
  FormLabel,
  FormGroup,
} from '@edx/paragon';

import { messages } from 'data/constants/app';
import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

const { csvUploadLabel, importBtnText } = messages.BulkManagementTab;

/**
 * <FileUploadForm />
 * File-type input wrapped with hidden control such that when a valid file is
 * added, it is automattically uploaded.
 */
export class FileUploadForm extends React.Component {
  constructor(props) {
    super(props);
    this.fileInputRef = React.createRef();
    this.handleClickImportGrades = this.handleClickImportGrades.bind(this);
    this.handleFileInputChange = this.handleFileInputChange.bind(this);
  }

  get fileInput() {
    return this.fileInputRef.current;
  }

  get formData() {
    const data = new FormData();
    data.append('csv', this.fileInput.files[0]);
    return data;
  }

  get hasFile() {
    return this.fileInput && this.fileInput.files[0];
  }

  handleClickImportGrades() {
    if (this.fileInput) { this.fileInput.click(); }
  }

  handleFileInputChange() {
    return this.hasFile && (
      this.props.submitFileUploadFormData(this.formData).then(
        () => { this.fileInput.value = null; },
      )
    );
  }

  render() {
    return (
      <>
        <Form action={this.props.gradeExportUrl} method="post">
          <FormGroup controlId="csv">
            <FormLabel>{csvUploadLabel}</FormLabel>
            <FormControl
              className="d-none"
              type="file"
              onChange={this.handleFileInputChange}
              ref={this.fileInputRef}
            />
          </FormGroup>
        </Form>

        <Button variant="primary" onClick={this.handleClickImportGrades}>
          {importBtnText}
        </Button>
      </>
    );
  }
}
FileUploadForm.propTypes = {
  // redux
  gradeExportUrl: PropTypes.string.isRequired,
  submitFileUploadFormData: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  gradeExportUrl: selectors.root.gradeExportUrl(state),
});

export const mapDispatchToProps = {
  submitFileUploadFormData: thunkActions.grades.submitFileUploadFormData,
};

export default connect(mapStateToProps, mapDispatchToProps)(FileUploadForm);
