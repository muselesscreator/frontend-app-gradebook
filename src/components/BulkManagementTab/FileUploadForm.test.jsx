/* eslint-disable import/no-named-as-default */
import React from 'react';
import { shallow, mount } from 'enzyme';
import {
  Button,
  Form,
  FormControl,
  FormGroup,
} from '@edx/paragon';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';
import * as appConstants from 'data/constants/app';

import { FileUploadForm, mapStateToProps, mapDispatchToProps } from './FileUploadForm';

const {
  messages: { BulkManagementTab: messages },
} = appConstants;

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    grades: {
      bulkImportError: jest.fn(state => ({ bulkImportError: state })),
    },
    root: {
      gradeExportUrl: jest.fn(state => ({ gradeExportUrl: state })),
    },
  },
}));
jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    grades: { submitFileUploadFormData: jest.fn() },
  },

}));
jest.mock('./BulkManagementAlerts', () => 'BulkManagementAlerts');
jest.mock('./ResultsSummary', () => 'ResultsSummary');

describe('FileUploadForm', () => {
  describe('component', () => {
    let props;
    let el;
    beforeEach(() => {
      props = {
        gradeExportUrl: 'fakeUrl',
        submitFileUploadFormData: jest.fn(),
      };
    });
    describe('snapshot', () => {
      const snapshotSegments = [
        'export form w/ alerts and file input',
        'import btn',
      ];
      test(`snapshot - loads ${snapshotSegments.join(', ')}`, () => {
        jest.mock('@edx/paragon', () => ({
          Button: () => 'Button',
          Form: () => 'Form',
          FormControl: () => 'FormControl',
          FormGroup: () => 'FormGroup',
        }));
        el = shallow(<FileUploadForm {...props} />);
        el.instance().handleFileInputChange = jest.fn().mockName('this.handleFileInputChange');
        el.instance().fileInputRef = jest.fn().mockName('this.fileInputRef');
        el.instance().handleClickImportGrades = jest.fn().mockName('this.handleClickImportGrades');
        el.instance().formatHistoryRow = jest.fn(entry => entry.originalFilename);
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('render', () => {
      beforeEach(() => {
        el = mount(<FileUploadForm {...props} />);
      });
      describe('alert form', () => {
        let form;
        beforeEach(() => {
          form = el.find(Form);
        });
        test('post action points to gradeExportUrl', () => {
          expect(form.props().action).toEqual(props.gradeExportUrl);
          expect(form.props().method).toEqual('post');
        });
        describe('file input', () => {
          let formGroup;
          beforeEach(() => {
            formGroup = el.find(FormGroup);
          });
          test('group with controlId="csv"', () => {
            expect(formGroup.props().controlId).toEqual('csv');
          });
          test('file control with onChange from handleFileInputChange', () => {
            const control = el.find(FormControl);
            expect(
              control.props().onChange,
            ).toEqual(el.instance().handleFileInputChange);
          });
          test('fileInputRef points to control', () => {
            expect(el.find(FormControl).getElement().ref).toBe(el.instance().fileInputRef);
          });
        });
      });
      describe('import button', () => {
        let btn;
        beforeEach(() => {
          btn = el.find(Button);
        });
        test('handleClickImportGrade on click', () => {
          expect(btn.props().onClick).toEqual(el.instance().handleClickImportGrades);
        });
        test('text from messages.importBtn', () => {
          expect(btn.children().text()).toEqual(messages.importBtnText);
        });
      });
    });
    describe('fileInput helper', () => {
      test('links to fileInputRef.current', () => {
        el = mount(<FileUploadForm {...props} />);
        const ref = 'a-fake-ref';
        el.instance().fileInputRef = { current: ref };
        expect(el.instance().fileInput).toEqual(ref);
      });
    });
    describe('behavior', () => {
      let fileInput;
      beforeEach(() => {
        el = mount(<FileUploadForm {...props} />);
        fileInput = jest.spyOn(el.instance(), 'fileInput', 'get');
      });
      describe('handleFileInputChange', () => {
        it('does nothing (does not fail) if fileInput has not loaded', () => {
          fileInput.mockReturnValue(null);
          el.instance().handleClickImportGrades();
        });
        it('calls fileInput.click if is loaded', () => {
          const click = jest.fn();
          fileInput.mockReturnValue({ click });
          el.instance().handleClickImportGrades();
          expect(click).toHaveBeenCalled();
        });
      });
      describe('handleClickImportGrades', () => {
        it('does nothing if file input has not loaded with files', () => {
          fileInput.mockReturnValue(null);
          el.instance().handleFileInputChange();
          expect(props.submitFileUploadFormData).not.toHaveBeenCalled();
          fileInput.mockReturnValue({ files: [] });
          el.instance().handleFileInputChange();
          expect(props.submitFileUploadFormData).not.toHaveBeenCalled();
        });
        it('calls submitFileUploadFormData and then clears fileInput if has files', () => {
          fileInput.mockReturnValue({ files: ['some', 'files'], value: 'a value' });
          const formData = { fake: 'form data' };
          jest.spyOn(el.instance(), 'formData', 'get').mockReturnValue(formData);
          const submit = jest.fn(() => ({ then: (thenCB) => { thenCB(); } }));
          el.setProps({
            submitFileUploadFormData: submit,
          });
          el.instance().handleFileInputChange();
          expect(submit).toHaveBeenCalledWith(formData);
          expect(el.instance().fileInput.value).toEqual(null);
        });
      });
      describe('formData', () => {
        test('returns FormData object with csv value from fileInput.files[0]', () => {
          const file = { a: 'fake file' };
          const value = 'aValue';
          fileInput.mockReturnValue({ files: [file], value });
          const expected = new FormData();
          expected.append('csv', file);
          expect([...el.instance().formData.entries()]).toEqual([...expected.entries()]);
        });
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { a: 'simple', test: 'state' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('gradeExportUrl from root.gradeExportUrl', () => {
      expect(mapped.gradeExportUrl).toEqual(selectors.root.gradeExportUrl(testState));
    });
  });

  describe('mapDispatchToProps', () => {
    test('submitFileUploadFormData from thunkActions.grades', () => {
      expect(
        mapDispatchToProps.submitFileUploadFormData,
      ).toEqual(thunkActions.grades.submitFileUploadFormData);
    });
  });
});
