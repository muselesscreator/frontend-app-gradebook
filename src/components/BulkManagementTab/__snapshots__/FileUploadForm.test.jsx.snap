// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploadForm component snapshot snapshot - loads export form w/ alerts and file input, import btn 1`] = `
<React.Fragment>
  <Form
    action="fakeUrl"
    inline={false}
    method="post"
  >
    <FormGroup
      as="div"
      controlId="csv"
      isInvalid={false}
      isValid={false}
    >
      <ForwardRef
        as="input"
        className="d-none"
        label="Upload Grade CSV"
        onChange={[MockFunction this.handleFileInputChange]}
        plaintext={false}
        type="file"
      />
    </FormGroup>
  </Form>
  <ForwardRef
    active={false}
    disabled={false}
    onClick={[MockFunction this.handleClickImportGrades]}
    variant="primary"
  >
    Import Grades
  </ForwardRef>
</React.Fragment>
`;
